enhancements/authentication/oauth-resource-storage.md

```diff
@@ -0,0 +1,207 @@
+
+A side-effect of non-sensitive object names is that these can be exposed via the
+API to other parties than the owner of the token in the session using the token. 
+This is helpful to create an API that allows a user to list his token objects, 
+e.g. to implement a feature to log out all "other" tokens in a REST-ful way.
+
+### Goals
+
+- Change the `oauthaccesstokens` and `oauthauthorizetokens` storage format and 
+  schema such that the etcd keys alone (which are not encrypted when encryption 
+  at REST is enabled) are not sensitive.
+- Make a "log out all other sessions" feature possible in the future.
+
+### Non-Goals
+
+- Design the "log out all other sessions" feature.
+
+## Proposal
+
+We prefix oauth access and oauth authorize token names with a prefix:
+
+- no prefix for old tokens,
+- `"sha256:"` for new tokens. The name is the sha256 hash of the actual token (not decoded, but plain base64)
+  encoded as base64.
+
+In v4.(n-1) and in v4.n:
+- we add support to verify sha256 tokens.
+
+In v4.n:
+- we create tokens of the new format.
+
+In v4.(n+1):
+ 
+- we remove the support to verify old tokens from the kube-apiserver authenticator.
+- we might add a feature to allow logouts of all other session.
+
+### User Stories [optional]
+
+#### Story 1
+
+As an attacker who got access to an etcd backup, but not the encryption keys I do 
+**not** get access to oauth tokens.
+
+#### Story 2
+
+Using an old oc (pre v4.n-1) I want to log into a v4.n or v4.n-1 (after downgrade) 
+cluster.
+
+#### Story 3
+
+With a token created in v4.n-1 I want to access a v4.n or v4.n+1 cluster.
+
+#### Story 4
+
+With a token created in v4.n I want to access a v4.n-1 cluster after downgrade.
+
+#### Story 5
+
+With a token created in v4.n-1 I do not expect to access a v4.n+1 cluster, i.e. the token will be invalid and I have to relogin.
+
+#### Non-goal Story 6
+
+(should be feasible, but a non-goal to design & implement) As a user I can use the API via oc or console to log out other sessions. 
+
+#### Non-goal Story 7
+
+(should be feasible, but a non-goal to design & implement) As a user I can list my other oauth sessions, including information like client name and expiration.
+
+### Implementation Details/Notes/Constraints
+
+#### Authenticator
+
+The oauth authenticator in kube-apiserver today does a GET look-up of the passed
+oauth token against the API server (with a caching layer in-front).
+
+In v4.n and v4.n-1, it will be extended:
+
+- if the passed token in the request is prefixed with `sha256:` the token value
+  is hashed and the hashed value is looked up to find the oauthaccesstoken 
+  object.
+  
+#### oauth-server
+
+The oauth-server creates access and authorize tokens. In v4.n it will hash the token value before using
+it as an object name:
+
+- `oauth-server/pkg/osinserver/registrystorage`: 
+  - in `LoadAuthorize`, `SaveAuthorize`,
+ `  RemoveAuthorize`, `SaveAccess`, `LoadAccess` will hash the passed token before using
```

Comment:
These should accomodate the `sha256:` prefix idea as well for cases where authorization token comes from an older version.

While we don't necessarily use `LoadAccess`, it'd be nice to have it fixed in case we decide to change that in the future.

Reply:
clarified.